File: scripts/3d/generate_dashboard_assets.py
"""Rebuild the Dashboard data-universe kit with Blender 4.x/5.x.

Run headlessly from the project root:
  blender --background --factory-startup --python scripts/3d/generate_dashboard_assets.py

The generated geometry is project-authored, texture-free, Y-up at runtime, and
distributed under the repository MIT license. Blender is a build-time tool only.
"""

import argparse
import bpy
from pathlib import Path
import sys

ROOT = Path(__file__).resolve().parents[2]
DEFAULT_OUTPUT = ROOT / "src" / "assets" / "3d" / "dashboard-data-universe.glb"
DEFAULT_BLEND_OUTPUT = ROOT / "scripts" / "3d" / "generated" / "dashboard-data-universe.blend"


def parse_args():
    parser = argparse.ArgumentParser(description="Generate the Dashboard data-universe GLB kit")
    parser.add_argument("--output", type=Path, default=DEFAULT_OUTPUT)
    parser.add_argument("--blend-output", type=Path, default=DEFAULT_BLEND_OUTPUT)
    argv = sys.argv[sys.argv.index("--") + 1 :] if "--" in sys.argv else []
    return parser.parse_args(argv)


def make_material(name, base, emission, metallic):
    material = bpy.data.materials.new(name)
    material.diffuse_color = (*base, 1.0)
    material.use_nodes = True
    shader = material.node_tree.nodes.get("Principled BSDF")
    shader.inputs.get("Base Color").default_value = (*base, 1.0)
    shader.inputs.get("Metallic").default_value = metallic
    shader.inputs.get("Roughness").default_value = 0.24
    emission_input = shader.inputs.get("Emission Color") or shader.inputs.get("Emission")
    if emission_input:
        emission_input.default_value = (*emission, 1.0)
    emission_strength = shader.inputs.get("Emission Strength")
    if emission_strength:
        emission_strength.default_value = 1.2
    return material


def add_cylinder(name, radius, depth, runtime_y, material):
    # Blender is Z-up; the glTF exporter maps this Z coordinate to runtime Y.
    bpy.ops.mesh.primitive_cylinder_add(vertices=48, radius=radius, depth=depth, location=(0, 0, runtime_y))
    obj = bpy.context.object
    obj.name = name
    obj.data.materials.append(material)
    return obj


def add_torus(name, major, minor, runtime_y, material, tilt_x=0.0, tilt_y=0.0):
    # A Blender torus already lies in the XY plane with a Z normal. After Y-up
    # export that is the horizontal runtime orientation, so the primary ring
    # needs no 90-degree correction.
    bpy.ops.mesh.primitive_torus_add(
        major_radius=major,
        minor_radius=minor,
        major_segments=96,
        minor_segments=8,
        location=(0, 0, runtime_y),
        rotation=(tilt_x, tilt_y, 0),
    )
    obj = bpy.context.object
    obj.name = name
    obj.data.materials.append(material)
    return obj


ARGS = parse_args()
OUTPUT = ARGS.output.resolve()
BLEND_OUTPUT = ARGS.blend_output.resolve()

bpy.ops.object.select_all(action="SELECT")
bpy.ops.object.delete(use_global=False)

cyan = make_material("CyanMetal", (0.043, 0.31, 0.47), (0.055, 0.65, 0.91), 0.78)
violet = make_material("VioletMetal", (0.15, 0.086, 0.36), (0.55, 0.36, 0.96), 0.62)
core_mat = make_material("CoreCrystal", (0.55, 0.91, 1.0), (0.13, 0.83, 0.87), 0.26)

add_cylinder("AssetPlatform", 1.25, 0.16, -0.12, cyan)
add_cylinder("AssetLowerDeck", 1.55, 0.08, -0.23, violet)
add_torus("AssetRingPrimary", 1.18, 0.028, -0.01, cyan)
add_torus("AssetRingSecondary", 1.55, 0.018, -0.17, violet, 0.12, 0.18)
bpy.ops.mesh.primitive_ico_sphere_add(subdivisions=2, radius=0.34, location=(0, 0, 0.36))
bpy.context.object.name = "AssetCoreShell"
bpy.context.object.data.materials.append(core_mat)

OUTPUT.parent.mkdir(parents=True, exist_ok=True)
BLEND_OUTPUT.parent.mkdir(parents=True, exist_ok=True)
bpy.ops.wm.save_as_mainfile(filepath=str(BLEND_OUTPUT))
bpy.ops.export_scene.gltf(
    filepath=str(OUTPUT),
    export_format="GLB",
    export_yup=True,
    export_apply=True,
    export_materials="EXPORT",
    export_cameras=False,
    export_lights=False,
)
print(f"Blender version: {bpy.app.version_string}")
print(f"Saved source: {BLEND_OUTPUT} ({BLEND_OUTPUT.stat().st_size} bytes)")
print(f"Exported runtime asset: {OUTPUT} ({OUTPUT.stat().st_size} bytes)")

File: scripts/3d/generate-dashboard-assets.mjs
import fs from 'node:fs/promises';
import path from 'node:path';
import { Buffer } from 'node:buffer';
import process from 'node:process';
import { fileURLToPath } from 'node:url';
import * as THREE from 'three';
import { GLTFExporter } from 'three/addons/exporters/GLTFExporter.js';

if (!globalThis.FileReader) {
  globalThis.FileReader = class FileReader {
    readAsArrayBuffer(blob) {
      blob.arrayBuffer().then((result) => {
        this.result = result;
        this.onloadend?.({ target: this });
      });
    }

    readAsDataURL(blob) {
      blob.arrayBuffer().then((result) => {
        this.result = `data:${blob.type};base64,${Buffer.from(result).toString('base64')}`;
        this.onloadend?.({ target: this });
      });
    }
  };
}

const scriptDir = path.dirname(fileURLToPath(import.meta.url));
const projectDir = path.resolve(scriptDir, '..', '..');
const output = path.join(projectDir, 'src', 'assets', '3d', 'dashboard-data-universe.glb');

const root = new THREE.Group();
root.name = 'DashboardDataUniverseKit';

function material(name, color, emissive, metalness = 0.5) {
  const value = new THREE.MeshStandardMaterial({
    name,
    color,
    emissive,
    emissiveIntensity: 1.2,
    roughness: 0.24,
    metalness,
  });
  return value;
}

const cyanMetal = material('CyanMetal', 0x0b4f78, 0x0ea5e9, 0.78);
const violetMetal = material('VioletMetal', 0x26165c, 0x8b5cf6, 0.62);
const coreMaterial = material('CoreCrystal', 0x8be8ff, 0x22d3ee, 0.26);

const platform = new THREE.Mesh(new THREE.CylinderGeometry(1.12, 1.34, 0.16, 48), cyanMetal);
platform.name = 'AssetPlatform';
platform.position.y = -0.12;
root.add(platform);

const lowerDeck = new THREE.Mesh(new THREE.CylinderGeometry(1.42, 1.68, 0.08, 48), violetMetal);
lowerDeck.name = 'AssetLowerDeck';
lowerDeck.position.y = -0.23;
root.add(lowerDeck);

const core = new THREE.Mesh(new THREE.OctahedronGeometry(0.34, 2), coreMaterial);
core.name = 'AssetCoreShell';
core.position.y = 0.36;
root.add(core);

const ringPrimary = new THREE.Mesh(new THREE.TorusGeometry(1.18, 0.028, 8, 96), cyanMetal);
ringPrimary.name = 'AssetRingPrimary';
ringPrimary.rotation.x = Math.PI / 2;
ringPrimary.position.y = -0.01;
root.add(ringPrimary);

const ringSecondary = new THREE.Mesh(new THREE.TorusGeometry(1.55, 0.018, 8, 96), violetMetal);
ringSecondary.name = 'AssetRingSecondary';
ringSecondary.rotation.set(Math.PI / 2, 0.12, 0.18);
ringSecondary.position.y = -0.17;
root.add(ringSecondary);

root.traverse((object) => {
  if (!object.isMesh) return;
  object.geometry.computeVertexNormals();
});

const exporter = new GLTFExporter();
const binary = await exporter.parseAsync(root, {
  binary: true,
  onlyVisible: true,
  truncateDrawRange: true,
});
await fs.mkdir(path.dirname(output), { recursive: true });
await fs.writeFile(output, Buffer.from(binary));
const stats = await fs.stat(output);
process.stdout.write(`Generated ${path.relative(projectDir, output)} (${stats.size} bytes)\n`);
